core/corehttp/gateway_handler.go

```diff
@@ -273,116 +275,87 @@ func (i *gatewayHandler) postHandler(w http.ResponseWriter, r *http.Request) {
 	http.Redirect(w, r, ipfsPathPrefix+k.String(), http.StatusCreated)
 }
 
-func (i *gatewayHandler) putEmptyDirHandler(w http.ResponseWriter, r *http.Request) {
-	newnode := uio.NewEmptyDirectory()
-
-	key, err := i.node.DAG.Add(newnode)
-	if err != nil {
-		webError(w, "Could not recursively add new node", err, http.StatusInternalServerError)
+func (i *gatewayHandler) putHandler(w http.ResponseWriter, r *http.Request) {
+	rootPath := path.Path(r.URL.Path)
+	if err := rootPath.IsValid(); err != nil {
+		webError(w, "putHandler: ipfs path not valid", err, http.StatusBadRequest)
 		return
 	}
 
-	i.addUserHeaders(w) // ok, _now_ write user's headers.
-	w.Header().Set("IPFS-Hash", key.String())
-	http.Redirect(w, r, ipfsPathPrefix+key.String()+"/", http.StatusCreated)
-}
-
-func (i *gatewayHandler) putHandler(w http.ResponseWriter, r *http.Request) {
-	// TODO(cryptix): either ask mildred about the flow of this or rewrite it
-	webErrorWithCode(w, "Sorry, PUT is bugged right now, closing request", errors.New("handler disabled"), http.StatusInternalServerError)
-	return
-	urlPath := r.URL.Path
-	pathext := urlPath[5:]
-	var err error
-	if urlPath == ipfsPathPrefix+"QmUNLLsPACCz1vLxQVkXqqLX5R1X345qqfHbsf67hvA3Nn/" {
-		i.putEmptyDirHandler(w, r)
+	rsegs := rootPath.Segments()
+	if rsegs[0] == ipnsPathPrefix {
+		webError(w, "putHandler: updating named entries not supported", errors.New("WritableGateway: ipns put not supported"), http.StatusBadRequest)
 		return
 	}
 
 	var newnode *dag.Node
-	if pathext[len(pathext)-1] == '/' {
+	if rsegs[len(rsegs)-1] == "QmUNLLsPACCz1vLxQVkXqqLX5R1X345qqfHbsf67hvA3Nn" {
 		newnode = uio.NewEmptyDirectory()
 	} else {
+		var err error
 		newnode, err = i.newDagFromReader(r.Body)
```

Comment:
in these cases i prefer explicit rebinding as opposed to something that might lead to accidental shadowing:
```
dfr, err := i.newDagFromReader(r.Body)
if err != nil {
....
}

newnode = dfr
```

I personally find it cleaner and easier to reason about.